lib/rules/no-misleading-character-class.js

```diff
@@ -181,6 +195,59 @@ module.exports = {
         const sourceCode = context.sourceCode;
         const parser = new RegExpParser();
 
+        /**
+         * Determines what offset to start a granular report, if it's straightforward to do so.
+         * @param {string} char Individual character being reported on.
+         * @param {Node} node Parent string node to report within.
+         * @see https://github.com/eslint/eslint/pull/17515#issuecomment-1707697186s
+         * @returns {Object | null} Report offset, or null to report on the whole node.
+         */
+        function getReportStartOffset(char, node) {
+            if (node.type !== "Literal" && (node.type !== "TemplateLiteral" || node.expressions.length)) {
+                return null;
+            }
+
+            const offsetStart = node.raw.slice(0, char.start + char.raw.length + 2).indexOf(char.raw, char.start - 2);
+
+            return offsetStart === -1 ? null : offsetStart;
+        }
+
+        /**
+         * Generates a granular loc for context.report.
+         * @param {string} char Individual character being reported on.
+         * @param {string} kind What kind of character report is being made.
+         * @param {Node} node Parent string node to report within.
+         * @param {number} reportStartOffset Starting offset to report within the node.
+         * @returns {Object} Granular loc for context.report.
+         * @see https://github.com/eslint/eslint/pull/17515#issuecomment-1707697186s
+         */
+        function generateReportLocation(char, kind, node, reportStartOffset) {
```

Comment:
I do not like how hardcode-y this function turned out 😕 but it's the cleanest I could figure out. Is there some better way of doing things?

Earlier in the implementation I'd had a `switch (kind)` with logic for each possible kind. Each mostly either subtracted `1` or `2`. I ended up mostly deduplicating then.